crates/rpc/rpc/src/eth/filter.rs

```diff
@@ -55,6 +56,237 @@ where
+            !headers.is_empty()
+        {
+            Self::Cached(CachedMode { filter_inner, headers, current_idx: 0 })
+        } else {
+            Self::Range(RangeBlockMode {
+                filter_inner,
+                block_range_iter: BlockRangeInclusiveIter::new(
+                    from_block..=to_block,
+                    max_headers_range,
+                ),
+                current_headers: None,
+                current_header_idx: 0,
+            })
+        }
+    }
+
+    /// Gets the next (receipts, `maybe_block`, header, `block_hash`) tuple.
+    async fn next(
+        &mut self,
+    ) -> Result<
+        Option<(
+            Arc<Vec<ProviderReceipt<Eth::Provider>>>,
+            Option<Arc<reth_primitives_traits::RecoveredBlock<ProviderBlock<Eth::Provider>>>>,
+            <Eth::Provider as HeaderProvider>::Header,
+            B256,
+        )>,
+        EthFilterError,
+    > {
+        match self {
+            Self::Cached(cached) => cached.next().await,
+            Self::Range(range) => range.next().await,
+        }
+    }
+}
+
+impl<
+        'a,
+        Eth: RpcNodeCoreExt<Provider: BlockIdReader, Pool: TransactionPool> + EthApiTypes + 'static,
+    > CachedMode<'a, Eth>
+{
+    async fn next(
+        &mut self,
+    ) -> Result<
+        Option<(
+            Arc<Vec<ProviderReceipt<Eth::Provider>>>,
+            Option<Arc<reth_primitives_traits::RecoveredBlock<ProviderBlock<Eth::Provider>>>>,
+            <Eth::Provider as HeaderProvider>::Header,
+            B256,
+        )>,
+        EthFilterError,
+    > {
+        loop {
+            if self.current_idx >= self.headers.len() {
+                return Ok(None);
+            }
+
+            let (_idx, header) = &self.headers[self.current_idx];
+            let header_idx = self.current_idx; // save index before incrementing
+            self.current_idx += 1;
+
+            // calculate block hash
+            let block_hash = if header_idx + 1 < self.headers.len() {
+                // use parent hash of next header to get current header's hash
+                self.headers[header_idx + 1].1.parent_hash()
+            } else {
+                // last header, need to fetch hash from provider
+                self.filter_inner
+                    .provider()
+                    .block_hash(header.number())?
+                    .ok_or_else(|| ProviderError::HeaderNotFound(header.number().into()))?
+            };
+
+            // try cache first for performance
+            if let Some((receipts, maybe_block)) =
+                self.filter_inner.eth_cache().get_receipts_and_maybe_block(block_hash).await?
+            {
+                return Ok(Some((receipts, maybe_block, header.clone(), block_hash)));
+            }
+
+            // cache miss - fallback to storage to ensure correctness
+            if let Some(receipts) =
+                self.filter_inner.provider().receipts_by_block(block_hash.into())?
```

Comment:
we have this fn

https://github.com/paradigmxyz/reth/blob/badbe3d81d9f27d9999e9419bede04359c20b9cb/crates/storage/provider/src/providers/database/mod.rs#L515-L518

which I think we could adapt to return Vec of Vec for blocks as well this would likely be more efficient for entire block ranges

Reply:
makes sense, will prepare a separate PR for this

Reply:
proposed here https://github.com/paradigmxyz/reth/pull/16449